tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -126,134 +132,305 @@ void validate(
 }
 } // namespace
 
-TEST_P(DistributedTransformerTest, MLP_Layer) {
-  auto [use_aten_matmul, dtype] = GetParam();
-  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
-  FusionGuard fg(fusion.get());
-  auto mesh = DeviceMesh::createForNumDevices(communicator_->size());
-  at::ScalarType at_dtype = data_type_to_aten(dtype);
+std::vector<at::Tensor> reference_mlp(
+    at::Tensor x,
+    at::Tensor w0,
+    at::Tensor b0,
+    at::Tensor w1,
+    at::Tensor b1,
+    at::ScalarType at_dtype) {
+  at::manual_seed(0);
+  auto linear1 = at::matmul(x, w0).add(b0).to(at::kFloat);
+  auto gelu = at::gelu(linear1, "tanh");
+  auto linear2 = at::matmul(gelu.to(at_dtype), w1).add(b1).to(at::kFloat);
+  auto dropout = at::dropout(linear2, kDropoutProb, true);
+  return {linear1, gelu, linear2, dropout};
+}
 
-  TensorView* x = makeContigTensor(2, dtype);
-  TensorView* w0 = makeContigTensor(3, dtype);
-  TensorView* b0 = makeContigTensor(2, dtype);
-  TensorView* w1 = makeContigTensor(3, dtype);
-  TensorView* b1 = makeContigTensor(1, dtype);
+std::vector<at::Tensor> reference_mlp_backwards(
+    at::Tensor grad,
+    at::Tensor x,
+    at::Tensor mask,
+    at::Tensor w0,
+    at::Tensor b0,
+    at::Tensor w1,
+    at::ScalarType at_dtype) {
+  // recompute activations
+  auto linear0 = at::matmul(x, w0).add(b0).to(at::kFloat);
+  auto gelu = at::gelu(linear0, "tanh");
 
-  fusion->addInput(x);
-  fusion->addInput(w0);
-  fusion->addInput(b0);
-  fusion->addInput(w1);
-  fusion->addInput(b1);
+  // backwards pass
+  auto dropout_grad =
+      at::native_dropout_backward(grad, mask, 1.0 / (1.0 - kDropoutProb));
+  auto dropout_grad_q = dropout_grad.to(at_dtype);
+  auto matmul1_grad = at::matmul(dropout_grad_q, w1.transpose(0, 1));
+  auto matmul1_grad_w =
+      at::matmul(dropout_grad_q.transpose(0, 1), gelu.to(at_dtype))
+          .transpose(0, 1);
+  auto matmul1_grad_b = at::sum(dropout_grad, {0});
+  auto gelu_grad =
+      at::gelu_backward(matmul1_grad.to(at::kFloat), linear0, "tanh");
+  auto gelu_grad_q = gelu_grad.to(at_dtype);
+  auto matmul0_grad_b = at::sum(gelu_grad, {0});
+  auto matmul0_grad = at::matmul(gelu_grad_q, w0.transpose(0, 1));
+  auto matmul0_grad_w =
+      at::matmul(gelu_grad_q.transpose(0, 1), x).transpose(0, 1);
 
+  std::vector<at::Tensor> grads = {
+      dropout_grad,
+      matmul1_grad_w,
+      matmul1_grad_b,
+      gelu_grad,
+      matmul0_grad_w,
+      matmul0_grad_b,
+      matmul0_grad};
+  return grads;
+}
+
+std::vector<TensorView*> mlp(
```

Comment:
This is almost a copy and paste of the nvfuser mlp definition BUT the option to have matmuls decomposed into bcast+mul-sum is removed. If we want to keep those tests I can add back